src/Collections-Weak/WeakValueDictionary.class.st

```diff
@@ -16,10 +16,32 @@ Class {
 
 { #category : #adding }
 WeakValueDictionary >> add: anAssociation [
-	self at: anAssociation key put: anAssociation value.
+	"Set the value at key to be anObject.  If key is not found, create a new
+	entry for key and set is value to anObject. Answer anObject."
+
+	| index |
+	index := self findElementOrNil: anAssociation key.
+	(array at: index)
+		ifNil: [ self
+				atNewIndex: index
+				put:
+					((WeakValueAssociation key: anAssociation key value: anAssociation value asCollectionElement)
+						originalAssociationClass: anAssociation class;
+						yourself) ]
+		ifNotNil: [ :element | element value: anAssociation value asSetElement ].
 	^ anAssociation
 ]
 
+{ #category : #adding }
+WeakValueDictionary >> associationAt: key ifAbsent: aBlock [
+	"Answer the association with the given key.
+	If the key is not found, return the result of evaluating aBlock."
+
+	^ (array at: (self findElementOrNil: key))
+		ifNil: [ aBlock value ]
+		ifNotNil: [ :assoc | assoc asOriginalAssociation ]
```

Comment:
I do not like the idea that dictionary should take care about original association.
We store #key and #value. The association is an implementation detail or suitable transfer object to deliver a pair to the dictionary. And I think it is better to keep it like that. Extra responsibility will complicate logic. 

You are trying to make Undeclared structure to be a weak. We can modify users to use normal  accessing methods of the dictionary (#at:/#at:put:) instead of #add:  and #associationsAt.
The #add: here was probably for optimization to avoid extra wrapper objects (associations) to store bindings. But I really doubt it is important here because Undeclared is normally empty. 

What you think to try another approach? I believe it will be simpler

Reply:
@dionisiydk Thinking more about it, indeed this should not be in WeakValueDictionary to keep the association. We should fix the problem with the value been nil, but not the problem of keeping the right association.

For Undeclared we should maybe create another class for it. I don't know if using #at: will be enough because I guess the system uses the kind of association to perform things. I will talk more with Marcus.